### PR TITLE
fix: placement check for N-term mods during beam search

### LINKED ISSUE #679: BUG: Casanovo can not predict N-term PTMs
## Describe the Issue
`Spec2Pep._finish_beams` has the two branches of its N-terminal modification placement check swapped, so the check is wrong for both tokenizer orientations:

```
# casanovo/denovo/model.py
n_terminal_pos = (
    0 if self.tokenizer.reverse else final_pos[has_n_term]
)
internal_mods = ~token_is_nterm[has_n_term, n_terminal_pos]
```
A reversed tokenizer reverses the token list, so the decoder emits the C-terminus first and a valid N-terminal modification is the last token generated, at `final_pos`. Index 0 is the C-terminal residue. The check requires the opposite in both branches. Because `ModelRunner.initialize_tokenizer` hardcodes `reverse=True`, every de novo run hits the wrong branch.

The effect is silent: every beam carrying an N-terminal modification is discarded, so affected runs report zero N-terminally modified peptides, and with `n_beams=1`, the affected spectra produce no identification at all. Nothing is logged.

This is present on main, dev, and released v5.2.0; it was introduced in #575.

## Steps To Reproduce
Reproduce end to end with any checkpoint whose vocabulary contains N-terminal modifications, e.g. the released orbitrap weights and the bundled sample data with `n_beams: 1`:

`casanovo sequence sample_data/sample_preprocessed_spectra.mgf --model casanovo_orbitrap_v5-2-0.ckpt --config config.yaml`

This returns exactly 0 / 121 sequences with an N-term PTM while there should be 7 / 128.

## Expected Behavior
A beam whose tokens are in the order the tokenizer produces should be kept, and only the reversed (invalid) order should be discarded and N-terminally modified peptides should appear in the output.

## Environment:
- OS:  Ubuntu 26.04 LTS
- Casanovo Version: 5.2.0 (also reproduced on main https://github.com/Noble-Lab/casanovo/commit/7112962a89680823eb47ebead676d0d4ed8997c9 and dev [a3fded4](https://github.com/Noble-Lab/casanovo/commit/a3fded4b77c515bd4d7d40c019bf01651b08a192))
- Hardware Used: GPU: NVIDIA GeForce RTX 4050 Laptop GPU, CUDA Version: 13.2 (driver 595.84)

## Additional Context
- The inverted branch also accepts beams with the modification at index 0. Those detokenize to a trailing modification (e.g. PTLDEK[Acetyl]-), which is not valid ProForma, so `on_predict_batch_end` raises an exception when it parses the sequence to compute the precursor m/z. This does not surface with the released orbitrap weights and the sample MGF since such beams do not win a spectrum but could happen with a different model and/or dataset.

### PR DESCRIPTION
**Pull request by PigeonMark**
### Summary
Fixes #679 

The check that decides where an N-terminal modification may sit in a decoded beam has its two branches swapped, so it is wrong for both tokenizer orientations. Since `ModelRunner.initialize_tokenizer` always builds the tokenizer with `reverse=True`, every de novo run takes the wrong branch and silently discards every beam carrying an N-terminal modification.

```
# casanovo/denovo/model.py, Spec2Pep._finish_beams
n_terminal_pos = (
    0 if self.tokenizer.reverse else final_pos[has_n_term]
)
internal_mods = ~token_is_nterm[has_n_term, n_terminal_pos]
```

A reversed tokenizer reverses the token list, so the decoder emits the C-terminus first and a valid N-terminal modification is the last token generated, at final_pos. Index 0 is the C-terminal residue. The condition asks for the opposite in both branches.

Confirmed against the tokenizer itself (depthcharge 0.4.10):

```
>>> PeptideTokenizer.from_massivekb(reverse=True).split("[Acetyl]-PTLDEK")
['K', 'E', 'D', 'L', 'T', 'P', '[Acetyl]-']

```
This corroborates `on_predict_batch_end`, which tests `spec_match.sequence.startswith(mod)`. It already expects the decoded string to begin with the modification.

Introduced in #575. Present on main, dev, and released v5.2.0.

### Impact

Measured with casanovo_orbitrap_v5-2-0.ckpt with n_beams: 1

_sample_data/sample_preprocessed_spectra.mgf_ (128 spectra):
Before: 121 sequenced, 0 N-term-modified PSMs
After: 128 sequenced, 7 N-term-modified PSMs

_MassIVE-KB subset_ (19,506 spectra)
Before: 17,041 sequenced, 0 N-term-modified PSMs
After: 19,504 sequenced, 2,463 N-term-modified PSMs

There is a second, milder consequence: because the inverted check accepts beams with the modification at index 0, such a beam detokenizes to a trailing modification (PTLDEK[Acetyl]-), which is not valid ProForma and raises when `on_predict_batch_end` parses it to compute the precursor m/z. In practice this depends on the checkpoint (with the released orbitrap weights I could not get such a beam to win a spectrum, because they always lose on score first, even though 18,288 of 19,506 spectra do place an N-terminal token in a step-0 beam). I have gotten this exception with other checkpoints. The fix removes the possibility either way.

### Changes
_casanovo/denovo/model.py_
- Swapped the branches so the modification is required at the position the tokenizer actually emits it, with a comment stating the invariant.
- Widened the guard `if step > 1:` to `if step > 0:`. The > 1 bound was inherited from the pre-#575 implementation, which indexed `final_pos - 1`; the current code no longer does, since multiple modifications are caught by `num_modifications > 1`. The old bound let a two-token beam such as `["[Acetyl]-", "K"]` escape the check entirely when `min_peptide_len <= 2`. At step 1 with a stop token `final_pos = 0`, so no negative index is possible; step 0 stays excluded because a lone first token sits at `final_pos` under either orientation.

_tests/unit_tests/test_unit.py_
- `test_invalid_n_term_mod` was itself inverted and passed against the buggy code. It was added in #575 alongside the defect and encodes the same assumption, so it could not have caught this. 
                                                                                                                                                                                                                                     - The other four cases (no modification, the two multiple-modification cases, and the internal-modification case) are unchanged.
- Three new tests, each parameterised over reverse ∈ {True, False}, so the branches cannot be silently swapped again:
  - `test_n_term_mod_orientation_matches_tokenizer` takes the token order from `tokenizer.split()` rather than hardcoding it, asserts the beam is kept, and asserts the decoded sequence round-trips through `calculate_precursor_ions`, the call that runs in `on_predict_batch_end`. Deriving the order from the tokenizer means the test still pins down correct behaviour if the decoding direction ever changes.
  - `test_n_term_mod_wrong_end_discarded`: the opposite order is discarded, and the sequence it would have decoded to is confirmed unparseable.
  - `test_n_term_mod_short_beam_discarded` covers the step > 0 widening with a two-token beam.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **Bug Fixes**
  - Corrected validation and decoding of N-terminal modifications across tokenizer directions and short decoding beams.
  - Preserved valid modified peptides during greedy decoding.
  - Prevented malformed peptides from causing parsing failures during beam search.
  - Corrected the reported mass for carbamylation with ammonia loss at the N-terminus.
  - De novo sequencing results and related evaluation metrics may change.

- **Documentation**
  - Updated the unreleased changelog and version comparison links.

- **Tests**
  - Expanded coverage for valid and invalid N-terminal modifications and ProForma parsing failures.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

### CODE REVIEW

File: tests/unit_tests/test_unit.py [hunk 1]
```diff
@@ -2554,6 +2558,109 @@ def test_invalid_n_term_mod(tiny_config, reverse):
...
+    torch.Tensor of shape (1, max_peptide_len + 1)
+        The zero-padded beam.
+    """
+    tokens = torch.zeros(
+        1, max_peptide_len + 1, dtype=torch.int64, device=device
+    )
+    tokens[0, : len(aa_tokens)] = torch.tensor(
+        [tokenizer.index[aa] for aa in aa_tokens], device=device
+    )
+    return tokens
```
[the remark this answers is not in the document]
Comment (PigeonMark) [reply tests/unit_tests/test_unit.py:2587]: Adding it would break local convention: `tests/unit_tests/test_unit.py` has 0 return annotations across 79 function definitions, and `tests/conftest.py` has none either. CONTRIBUTING asks that contributions "integrate naturally with the existing codebase," so this is left it as-is.

**Comment by PigeonMark** [comment]
@bittremieux ready for review.
Given that Casanovo can't predict sequences with N-term PTMs at the moment, I would recommend a hotfix release

**Review by bittremieux — DISMISSED** [review]

**Review by bittremieux — APPROVED** [review]
